gobblin-modules/gobblin-service-kafka/src/main/java/org/apache/gobblin/service/StreamingKafkaSpecConsumer.java

```diff
@@ -111,13 +113,15 @@ public StreamingKafkaSpecConsumer(Config config, MutableJobCatalog jobCatalog) {
 
     try {
       Pair<SpecExecutor.Verb, Spec> specPair = _jobSpecQueue.take();
-      _metrics.specConsumerJobSpecDeq.mark();
+      int numSpecFetched = 0;
       do {
+        _metrics.specConsumerJobSpecDeq.mark();
+        numSpecFetched ++;
         changesSpecs.add(specPair);
 
         // if there are more elements then pass them along in this call
         specPair = _jobSpecQueue.poll();
-      } while (specPair != null);
+      } while (specPair != null && numSpecFetched < _jobSpecQueueSize);
```

Comment:
Yes, it's just to make it to be finite as I notice that during some peek time, we may end up with continually put elements into the queue and the loop takes a long time to finish and yet cause the job to delay. I picked this value because queue size is configurable and I think the max batch size(spec we fetched at one time) and the max queue capacity should be positively related.

Reply:
although likely not terribly detrimental to 'reuse' that configuration for this other purpose, I don't fully see the connection between the two.  while we wouldn't want to take just one element, as that could unblock only a single addition by the queue's producer, it may nonetheless be non-optimal to drain the entirety.  profiling would help chose the right relationship between producing and consuming, but maybe we'd begin between [.05, .5] of capacity.

again though, seemingly a smaller performance optimization we ought not to over-engineer, unless some presenting problem surfaces.  still, don't forget about the `.drainTo` method, to remove up to the desired amount all at once -
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/LinkedBlockingQueue.html#drainTo-java.util.Collection-int-